### ISSUE TITLE
0.31 RC1 crashes on very little load

### ISSUE THREAD

**Issue by doodlebro** [metabase/metabase#8868]
Hi, 

I am running metabase 0.31 RC1 on a Kubernetes cluster, but if I start to rifle through dashboards, collections, etc. the pod will crash. Logs don't leave much help, but I will leave the tail of the log just before the crash happened.

```11-06 22:16:45 DEBUG metabase.middleware :: GET /api/collection/14 200 (89 ms) (3 DB calls). Jetty threads: 8/50 (13 busy, 4 idle, 0 queued)
11-06 22:16:45 DEBUG metabase.middleware :: GET /api/collection/14/items 200 (17 ms) (8 DB calls). Jetty threads: 8/50 (13 busy, 4 idle, 0 queued)
11-06 22:16:53 INFO middleware.cache :: Query took 74100 ms to run; miminum for cache eligibility is 30000 ms
11-06 22:16:53 INFO middleware.cache :: Caching results for next time for query 💾
11-06 22:16:53 DEBUG metabase.middleware :: POST /api/card/434/query 200 (1 mins) (14 DB calls). Jetty threads: 8/50 (12 busy, 6 idle, 0 queued)
11-06 22:16:59 INFO middleware.cache :: Query took 79905 ms to run; miminum for cache eligibility is 30000 ms
11-06 22:16:59 INFO middleware.cache :: Caching results for next time for query 💾
11-06 22:16:59 DEBUG metabase.middleware :: POST /api/card/418/query 200 (1 mins) (14 DB calls). Jetty threads: 8/50 (11 busy, 7 idle, 0 queued)
11-06 22:17:25 INFO middleware.cache :: Query took 45497 ms to run; miminum for cache eligibility is 30000 ms
11-06 22:17:25 INFO middleware.cache :: Caching results for next time for query 💾
11-06 22:17:25 DEBUG metabase.middleware :: POST /api/card/434/query 200 (46 s) (13 DB calls). Jetty threads: 8/50 (10 busy, 8 idle, 0 queued)
11-06 22:17:33 INFO middleware.cache :: Query took 53792 ms to run; miminum for cache eligibility is 30000 ms
11-06 22:17:33 INFO middleware.cache :: Caching results for next time for query 💾
11-06 22:17:33 DEBUG metabase.middleware :: POST /api/card/418/query 200 (54 s) (13 DB calls). Jetty threads: 8/50 (9 busy, 9 idle, 0 queued)
11-06 22:17:33 INFO middleware.cache :: Query took 114088 ms to run; miminum for cache eligibility is 30000 ms
11-06 22:17:33 INFO middleware.cache :: Caching results for next time for query 💾
11-06 22:17:33 DEBUG metabase.middleware :: POST /api/card/527/query 200 (2 mins) (13 DB calls). Jetty threads: 8/50 (8 busy, 10 idle, 0 queued)
11-06 22:17:39 INFO api.card :: Question's average execution duration is 105663 ms; using 'magic' TTL of 63398 seconds 💾
11-06 22:17:39 INFO middleware.cache :: Returning cached results for query 💾
11-06 22:17:39 DEBUG metabase.middleware :: POST /api/card/527/query 200 (99 ms) (6 DB calls). Jetty threads: 8/50 (8 busy, 10 idle, 0 queued)
11-06 22:17:58 INFO middleware.cache :: Query took 78591 ms to run; miminum for cache eligibility is 30000 ms
11-06 22:17:58 INFO middleware.cache :: Caching results for next time for query 💾
11-06 22:17:58 DEBUG metabase.middleware :: POST /api/card/527/query 200 (1 mins) (13 DB calls). Jetty threads: 8/50 (7 busy, 11 idle, 0 queued)
11-06 22:18:09 INFO models.collection :: Moving Collection 19 and its descendants from / to /28/
11-06 22:18:09 DEBUG metabase.middleware :: PUT /api/collection/19 200 (68 ms) (13 DB calls). Jetty threads: 8/50 (7 busy, 11 idle, 0 queued)
11-06 22:18:09 DEBUG metabase.middleware :: GET /api/collection/root/items 200 (29 ms) (6 DB calls). Jetty threads: 8/50 (7 busy, 11 idle, 0 queued)
11-06 22:18:09 DEBUG metabase.middleware :: GET /api/collection/28 200 (7 ms) (3 DB calls). Jetty threads: 8/50 (7 busy, 11 idle, 0 queued)
11-06 22:18:09 DEBUG metabase.middleware :: GET /api/collection/28/items 200 (14 ms) (7 DB calls). Jetty threads: 8/50 (7 busy, 11 idle, 0 queued)
11-06 22:18:12 DEBUG metabase.middleware :: PUT /api/collection/19 200 (8 ms) (4 DB calls). Jetty threads: 8/50 (7 busy, 11 idle, 0 queued)
11-06 22:18:12 DEBUG metabase.middleware :: GET /api/collection/28/items 200 (14 ms) (7 DB calls). Jetty threads: 8/50 (7 busy, 11 idle, 0 queued)
11-06 22:18:18 DEBUG metabase.middleware :: GET /api/collection/root/items 200 (23 ms) (6 DB calls). Jetty threads: 8/50 (7 busy, 11 idle, 0 queued)
11-06 22:18:20 DEBUG metabase.middleware :: GET /api/collection/28 200 (7 ms) (3 DB calls). Jetty threads: 8/50 (7 busy, 11 idle, 0 queued)
11-06 22:18:21 DEBUG metabase.middleware :: GET /api/collection/19 200 (11 ms) (4 DB calls). Jetty threads: 8/50 (7 busy, 11 idle, 0 queued)
11-06 22:18:21 DEBUG metabase.middleware :: GET /api/collection/19/items 200 (15 ms) (8 DB calls). Jetty threads: 8/50 (7 busy, 11 idle, 0 queued)
11-06 22:18:27 DEBUG metabase.middleware :: GET /api/collection/root 200 (4 ms) (2 DB calls). Jetty threads: 8/50 (7 busy, 10 idle, 0 queued)
11-06 22:18:29 DEBUG metabase.middleware :: GET /api/collection/28 200 (8 ms) (3 DB calls). Jetty threads: 8/50 (7 busy, 10 idle, 0 queued)
11-06 22:18:34 DEBUG metabase.middleware :: GET /api/collection/root 200 (4 ms) (2 DB calls). Jetty threads: 8/50 (7 busy, 10 idle, 0 queued)
11-06 22:18:36 DEBUG metabase.middleware :: GET /api/collection/28 200 (8 ms) (3 DB calls). Jetty threads: 8/50 (7 busy, 10 idle, 0 queued)
11-06 22:18:40 DEBUG metabase.middleware :: GET /api/dashboard/24 200 (86 ms) (17 DB calls). Jetty threads: 8/50 (7 busy, 10 idle, 0 queued)
11-06 22:18:41 INFO api.card :: Question's average execution duration is 16697 ms; using 'magic' TTL of 10018 seconds 💾
11-06 22:18:41 INFO api.card :: Question's average execution duration is 63302 ms; using 'magic' TTL of 37981 seconds 💾
11-06 22:18:41 INFO api.card :: Question's average execution duration is 48295 ms; using 'magic' TTL of 28977 seconds 💾
11-06 22:18:41 INFO api.card :: Question's average execution duration is 34697 ms; using 'magic' TTL of 20818 seconds 💾
11-06 22:18:41 INFO api.card :: Question's average execution duration is 46793 ms; using 'magic' TTL of 28076 seconds 💾
11-06 22:18:41 INFO middleware.cache :: Query took 413 ms to run; miminum for cache eligibility is 30000 ms
11-06 22:18:41 DEBUG metabase.middleware :: POST /api/card/235/query 200 (527 ms) (11 DB calls). Jetty threads: 8/50 (12 busy, 5 idle, 0 queued)
11-06 22:18:42 INFO api.card :: Question's average execution duration is 41058 ms; using 'magic' TTL of 24635 seconds 💾
11-06 22:18:42 INFO middleware.cache :: Query took 499 ms to run; miminum for cache eligibility is 30000 ms
11-06 22:18:42 DEBUG metabase.middleware :: POST /api/card/238/query 200 (609 ms) (11 DB calls). Jetty threads: 8/50 (12 busy, 5 idle, 0 queued)
11-06 22:18:42 INFO api.card :: Question's average execution duration is 37404 ms; using 'magic' TTL of 22442 seconds 💾
```

Hoping to get this fixed before the full 0.31 release as the new features are really useful to our org!

**Comment by salsakran** [comment]
Does this happen every time? Do you have any monitoring on that instance you could share? 

Would really like to get to the bottom of this.

**Comment by doodlebro** [comment]
@salsakran: Yeah, I can get it to crash within 5-10 minutes of it coming up each time. If there is no user interaction, the instance actually stays up for quite awhile. Looks like the last time it crashed, there was no user interaction but it lasted 12 hours. The current instance has been up for 7 hours and is still up because I haven't logged in.

Forgive me for being a bit dense here, but what would shareable monitoring look like for my setup? Totally happy to get to the bottom of this, let me know what more I can provide!

**Comment by doodlebro** [comment]
I'm running Metabase on a Kubernetes pod with 2GB of allocated memory... I'm thinking that could be the problem so I upped it to 4GB... no crashes for 30 minutes of playing. I'll keep an eye on it and close this if I can't get it to reproduce with higher memory.

**Comment by camsaul** [comment]
@doodlebro have you configured any JVM memory flags? `-Xmx` in particular?

**Comment by senior** [comment]
@doodlebro If you are not explicitly setting a max heap, it will use the JVM defaults. If you have set it to allocate 2GB to the container, Metabase will have a 512MB max heap. When you bumped the container to 4GB, it would use a 1GB heap.

How does the 2GB vs. 4GB setting for the RC1 compare to your production settings for prior versions. Are you running on 4GB in production?

**Comment by doodlebro** [comment]
Production was also at 2GB, but I upped it to 4GB yesterday to see if there was any performance increase. Since our Production instance is such an old version (0.29), it makes sense that the memory usage is much lower. I did not notice any substantial performance increase or reduced wait times with this change on production, but I am leaving it at 4GB as I intend to upgrade to 0.31 when it releases. Production would very rarely have crashing at 2GB, but it was incredibly uncommon.

Thanks for the info on heap size. I don't think I need to make any adjustments since the crash has not come back, but I'll keep it in mind in case it does.
